libp2p/io/utils.py

```diff
@@ -0,0 +1,21 @@
+from libp2p.io.abc import Reader
+from libp2p.io.exceptions import IncompleteReadError
+
+DEFAULT_RETRY_READ_COUNT = 100
+
+
+async def read_exactly(
+    reader: Reader, n: int, retry_count: int = DEFAULT_RETRY_READ_COUNT
+) -> bytes:
+    """
+    NOTE: relying on exceptions to break out on erroneous conditions, like EOF
+    """
+    data = await reader.read(n)
+
+    for _ in range(retry_count):
+        if len(data) < n:
+            remaining = n - len(data)
+            data += await reader.read(remaining)
```

Comment:
Would we want to sleep for a while(e.g. 10 ms) in every loop, to yield the CPU and wait for a while for the data? Otherwise, it seems to me that this `retry_count` might be not so efficient. I checked the code in `asyncio.Reader` it seems in every call to `asyncio.Reader.read` there is a `wait_for_data`. However, in other implementations of `Reader`, `Reader.read` can possibly return right away if there is no data in buffer available.

Reply:
my thinking here is that we will yield the control flow at every `await` -- i could be wrong here but if this is in fact how python async works then looping around a bunch would imply there is not much else waiting on the event loop anyway... i figured it would become obvious if this function is a problem (e.g. it would look like blocking the event loop) at which point we can revisit this -- in the interest of time, i'd favor moving ahead and merging this for now

Reply:
Yeah, I don't have a perfect model of asyncio in my head, but `await` always yielding control flow is how I think of it. I don't really follow what alternative you're suggesting, though.

Reply:
the alternative would be to explicitly add something like

```python
for _ in range(retry_count):
    await asyncio.sleep(0.1)
    if len(data) < n:
        remaining = n - len(data)
        ...
```

so that we ensure to yield to the event loop on every loop iteration (... too many loops...).

something like `await asyncio.sleep(0)` is sometimes recommended to break up a chunk of pure compute so that you don't block the thread if there are other coroutines waiting to make progress. in this case i think we are fine like i said above bc we will yield on the `reader.read`.

there is the separate issue we are all touching on which is "does this function satisfy the semantics of what we want to call 'read_exactly'" and i agree its less clear that this is the case -- my intention was to just keep an eye on it :)

Reply:
I'm coming in blind here but wanted to add that `asyncio.StreamReader`s already has [a `readexactly` method](https://github.com/python/cpython/blob/master/Lib/asyncio/streams.py#L1056), it might be useful to see how asyncio does it.

@mhchia I think your concern is that non-asyncio implementations of `libp2p.io.abc.Reader` might return immediately if there is no data? So this coroutine could end up spinning forever as it calls `read()` and gets nothing and calls `read()` again? I think a reasonable way to fix that situation would be to say that `read()` must always block until it can return at least one byte, another way might be to add a method like `wait_for_data` which this method can use. But adding `asyncio.sleep()` is usually a sign that there's a better design to be found somewhere, in this case it would mean the client not only does useless work every 10ms (which would hurt the battery of whatever system it's running on) but it also assumes an artificial 5ms (on average) delay before it even notices incoming messages!

Reply:
> I think your concern is that non-asyncio implementations of libp2p.io.abc.Reader might return immediately if there is no data? So this coroutine could end up spinning forever as it calls read() and gets nothing and calls read() again?

Yeah exactly. Since there is nothing mentioned in the `Reader` interface about the behavior of `read`, I was thinking people are possibly not aware that, `readexactly` doesn't help if their implementation does not blocking-wait for the first byte. Probably a note or warning in this function will be helpful.

> I think a reasonable way to fix that situation would be to say that read() must always block until it can return at least one byte, another way might be to add a method like wait_for_data which this method can use.

This way makes a lot of sense:) 

> But adding asyncio.sleep() is usually a sign that there's a better design to be found somewhere, in this case it would mean the client not only does useless work every 10ms (which would impact the battery of whatever system it's running on) but it also assumes an artificial 5ms (on average) delay before it even notices incoming messages!

Good point. I think it might be better to leave it as is.